src/field/gen_fieldops.py

```diff
@@ -0,0 +1,211 @@
+        # array: dimensions of the field
+        self.dimensions = dimensions
+        # name of this field
+        self.name = name
+
+    @property
+    def passByReference(self):
+        """Returns "Type& name", except if field_type is BoutReal,
+        in which case just returns "Type name"
+
+        """
+        return "{self.field_type}{ref} {self.name}".format(
+            self=self, ref="&" if self.field_type != "BoutReal" else "")
+
+    @property
+    def passBoutRealPointer(self):
+        """Returns "BoutReal* name", except if field_type is BoutReal,
+        in which case just returns "BoutReal name"
+
+        Also adds "__restrict__" attribute if `for_gcc` is True
+        """
+        return "BoutReal {ref}{restrict} {self.name}".format(
+            self=self, ref="*" if self.field_type != "BoutReal" else "",
+            restrict="__restrict__" if for_gcc and self.field_type != "BoutReal" else "")
+
+    @property
+    def getPointerToData(self):
+        """Returns a pointer to the underlying data, e.g. `&Field3D[i]`,
+        except for plain BoutReals"""
+        if self.field_type == 'BoutReal':
+            return self.name
+        return "&{}[i]".format(self.name)
+
+    def getElement(self, data=True):
+        """Get an element in the field, either using C-loop style index (default),
+        or with an `Indices` object (if `data=False`)
+
+        """
+
+        if self.field_type == 'BoutReal':
+            return self.name
+        if data:
+            if self.field_type == 'Field2D':
+                index = "y + x*ny"
+            elif self.field_type == 'FieldPerp':
+                index = "z + x*nz"
+            elif self.field_type == 'Field3D':
+                index = "z + nz*(y + ny*x)"
+            else:
+                raise ValueError("Unexpected field_type")
+        else:
+            index = "i"
+        return "{name}[{index}]".format(name=self.name, index=index)
+
+    def __eq__(self, other):
+        try:
+            return self.field_type == other.field_type
+        except AttributeError:
+            return self.field_type == other
+
+    def __ne__(self, other):
+        return not (self == other)
+
+    def __repr__(self):
+        return "Field({}, {}, {})".format(self.field_type, self.dimensions, self.name)
+
+    def __str__(self):
+        return self.field_type
+
+
+# Declare what fields we currently support:
+# Field perp is currently missing
+field3D = Field('Field3D', ['x', 'y', 'z'])
+field2D = Field('Field2D', ['x', 'y'])
+boutreal = Field('BoutReal', [])
+fields = [field3D, field2D, boutreal]
+
+
+def returnType(f1, f2):
+    """Determine a suitable return type, by seeing which field is 'larger'.
+
+    """
+    if f1 == f2:
+        return copy(f1)
+    elif f1 == 'BoutReal':
+        return copy(f2)
+    elif f2 == 'BoutReal':
+        return copy(f1)
+    else:
+        return copy(field3D)
+
+
+if __name__ == "__main__":
+    for_gcc = True
+    print(header)
```

Comment:
We print to stdout and then redirect to file. Could this cause problems if something develops a warning/we want to provide updates on the process to the user via messages? Could we perhaps take an optional file to write to instead (e.g. using `sys.arg` or similar).

Reply:
That's sensible. The `for_gcc` argument could be passed in as well. I'm not entirely sure why I didn't do that already.